block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagRepresentationSyntax.scala

```diff
@@ -159,4 +159,23 @@ final class BlockDagRepresentationOps[F[_]](
       .flatMap(Stream.emits)
       .compile
       .to(Set)
+
+  def ancestors(blockHash: BlockHash, filterF: BlockHash => F[Boolean])(
+      implicit sync: Sync[F]
+  ): F[Set[BlockHash]] =
+    Stream
+      .unfoldEval(List(blockHash)) { lvl =>
+        val parents = lvl
+          .traverse(lookupUnsafe)
+          .flatMap(_.flatMap(_.parents).distinct.filterA(filterF))
+        parents.map(p => p.nonEmpty.guard[Option].as(p, p))
+      }
+      .flatMap(Stream.emits)
+      .compile
+      .to(Set)
+
+  def withAncestors(blockHash: BlockHash, filterF: BlockHash => F[Boolean])(
+      implicit sync: Sync[F]
+  ): F[Set[BlockHash]] =
+    ancestors(blockHash, filterF).map(_ + blockHash)
```

Comment:
If `ancestors` return Stream then filter can be be used on that stream directly without predicate function as a parameter. This is the point of Stream as being lazy and pull based so consumer can decide when to stop pulling.

```scala
def ancestors2(blockHash: BlockHash)(
    implicit sync: Sync[F]
): Stream[F, BlockHash] =
  Stream
    .unfoldEval(List(blockHash)) { lvl =>
      val parents = lvl
        .traverse(lookupUnsafe)
        .map(_.flatMap(_.parents).distinct)
      parents.map(p => p.nonEmpty.guard[Option].as(p, p))
    }
    .flatMap(Stream.emits)

def withAncestors(blockHash: BlockHash, filterF: BlockHash => F[Boolean])(
    implicit sync: Sync[F]
): F[Set[BlockHash]] =
  ancestors2(blockHash).evalFilter(filterF).compile.to(Set)
```

Reply:
Deduplication can be done on a Stream also, but this is not the issue here. What I mean is that caller can call `.compile.to(Set)` to get unique elements. There is not need to invent the same functionality that already exists on Stream like another way of filtering which means `withAncestors` is just code duplication.

Reply:
`What I mean is that caller can call .compile.to(Set) to get unique elements.` I understand that, I'm saying that if I would see 
```
def ancestors2(blockHash: BlockHash)(
    implicit sync: Sync[F]
): Stream[F, BlockHash] 
```
I would assume that I can just call `compile.toList` and thats it, I would not expect this list contain duplicate elements. I made it return Set to be specific, that this will explicitly return unique ancestors. So function user not introduce bug assuming stream does not contain duplicates.

Reply:
Deduplication must involve some state to keep track of previous elements. Here is some discussion.
https://gitter.im/functional-streams-for-scala/fs2?at=58c300d41465c46a56152134
With the link to this gist.
https://gist.github.com/zeryx/5d918a433dca167975ca047a7e1f4cc1
We can create little modified version with Ref which stores Set of previous hashes and only produce unique ones, similar to what you described.

The main problem is that now `ancestors` cannot be used lazily just because includes materialization of Set at the end and because of that also needs filter as a parameter.

Reply:
Ok, I will make it return stream with comment about duplicates

Reply:
From discussion on standup, we concluded that with filter directly on a stream it's difficult to take only non finalized blocks because some parents may not be included. So better to leave as is.